rslib/src/card_rendering/writer.rs

```diff
@@ -0,0 +1,210 @@
+// Copyright: Ankitects Pty Ltd and contributors
+// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
+
+use std::fmt::Write as _;
+
+use super::{CardNodes, Node, OtherTag, Tag, TtsTag};
+use crate::prelude::*;
+use crate::{
+    backend_proto as pb,
+    text::{decode_entities, strip_html_for_tts},
+};
+
+impl<'a> CardNodes<'a> {
+    pub(super) fn write_without_av_tags(&self) -> String {
+        AvStripper::new().write(self)
+    }
+
+    pub(super) fn write_and_extract_av_tags(
+        &self,
+        question_side: bool,
+        tr: &I18n,
+    ) -> (String, Vec<pb::AvTag>) {
+        let mut extractor = AvExtractor::new(question_side, tr);
+        (extractor.write(self), extractor.tags)
+    }
+}
+
+trait Write {
+    fn write<'iter, 'nodes: 'iter, T>(&mut self, nodes: T) -> String
+    where
+        T: IntoIterator<Item = &'iter Node<'nodes>>,
+    {
+        let mut buf = String::new();
```

Comment:
This is probably not ideal. Should we iterate the nodes in advance to find an approximate capacity first?

Reply:
I wouldn't bother - the rewriting case is not a particularly hot path, as IIRC, the most we're going to deal with is a few hundred requests at a time when showing browser rows.

The way a trait has been used here surprised me a bit at first, as each time a new tag type is added, the trait would need to be extended. But I see you've done that to be able to reuse the default methods, and it seems fine.